fixed the windows ctrl-c bug

xiaomaogy opened the pull request.

xiaomaogy commented:
This solves the long persisting problem on windows that when doing the Ctrl-C on Windows, the model doesn't gets saved. With this fix the cptk model file gets saved properly, and the .bytes file also gets generated correctly. 

The problem with this fix is that some InvalidArgumentError message will get printed which doesn't look so nice.  I don't know how to turn off properly yet.

xiaomaogy commented:
I was using the solution posted in this stackoverflow page to do the fix if you are interested. https://stackoverflow.com/questions/15457786/ctrl-c-crashes-python-after-importing-scipy-stats

vincentpierre commented:
I have not tested it.
`global_step` is now a property. I am not sure how I feel about this change.
There is a lot of duplicate code in `_win_handler` can we create a single method that will be used both in `_win_handler` and when there is a keyboard interrupt ?
The pytest broke, I think you need to add the win32 pip package to the setup.py requirements. Can you describe the InvalidArgumentError message you get ?

xiaomaogy commented:
@vincentpierre Thanks for the comment! 
1. I had to change the global_step to be a property, so that I can use this global_step in the _win_handler function when the ctrl-c event gets triggered. I'm not sure how to pass this global_step inside the function without making it a property.  
2. Regarding duplicate code: Will do. 
3. Regarding pytest. Will fix. 
4. Regarding InvalidArgumentError, what I get is message saying that the input shape doesn't match the required shape of some Tensor. I guess this is we stopped some processed when ctrl-c event gets triggered, and the data stops flowing in, but some other process are still running and detected that no data flows in. I will post the complete error message later.

xiaomaogy commented:
The InvalidArgumentError looks like this:

```
INFO:mlagents.trainers: test1-0: 3DBallLearning: Step: 1000. Mean Reward: 1.199. Std of Reward: 0.645. Training.
INFO:mlagents.trainers: test1-0: 3DBallLearning: Step: 2000. Mean Reward: 1.322. Std of Reward: 0.747. Training.
INFO:mlagents.trainers: test1-0: 3DBallLearning: Step: 3000. Mean Reward: 1.776. Std of Reward: 1.122. Training.
INFO:mlagents.trainers: test1-0: 3DBallLearning: Step: 4000. Mean Reward: 2.193. Std of Reward: 1.437. Training.
INFO:mlagents.trainers: test1-0: 3DBallLearning: Step: 5000. Mean Reward: 3.271. Std of Reward: 2.488. Training.
--------------------------Now saving model-------------------------
INFO:mlagents.envs:Learning was interrupted. Please wait while the graph is generated.
INFO:mlagents.envs:Saved Model
INFO:mlagents.trainers:List of nodes to export for brain :3DBallLearning
INFO:mlagents.trainers: is_continuous_control
INFO:mlagents.trainers: version_number
INFO:mlagents.trainers: memory_size
INFO:mlagents.trainers: action_output_shape
INFO:mlagents.trainers: action
INFO:mlagents.trainers: action_probs
INFO:mlagents.trainers: value_estimate
Process Process-1:
Traceback (most recent call last):
  File "C:\Users\xiaom\Anaconda3\lib\site-packages\tensorflow\python\client\session.py", line 1327, in _do_call
    return fn(*args)
  File "C:\Users\xiaom\Anaconda3\lib\site-packages\tensorflow\python\client\session.py", line 1310, in _run_fn
    self._extend_graph()
  File "C:\Users\xiaom\Anaconda3\lib\site-packages\tensorflow\python\client\session.py", line 1358, in _extend_graph
    graph_def.SerializeToString(), status)
  File "C:\Users\xiaom\Anaconda3\lib\site-packages\tensorflow\python\framework\errors_impl.py", line 516, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.InvalidArgumentError: NodeDef expected inputs 'int32_ref, int32' do not match 0 inputs specified; Op<name=Assign; signature=ref:Ref(T), value:T -> output_ref:Ref(T); attr=T:type; attr=validate_shape:bool,default=true; attr=use_locking:bool,default=true; allows_uninitialized_input=true>; NodeDef: global_step/Assign_1 = Assign[T=DT_INT32, _class=["loc:@global_step"], use_locking=true, validate_shape=true]()

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\xiaom\Anaconda3\lib\multiprocessing\process.py", line 258, in _bootstrap
    self.run()
  File "C:\Users\xiaom\Anaconda3\lib\multiprocessing\process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "c:\users\xiaom\desktop\ml-agents\ml-agents\mlagents\trainers\learn.py", line 52, in run_training
    tc.start_learning()
  File "c:\users\xiaom\desktop\ml-agents\ml-agents\mlagents\trainers\trainer_controller.py", line 374, in start_learning
    trainer.update_policy()
  File "c:\users\xiaom\desktop\ml-agents\ml-agents\mlagents\trainers\ppo\trainer.py", line 343, in update_policy
    run_out = self.policy.update(buffer.make_mini_batch(start, end), n_sequences)
  File "c:\users\xiaom\desktop\ml-agents\ml-agents\mlagents\trainers\ppo\policy.py", line 143, in update
    run_out = self._execute_model(feed_dict, self.update_dict)
  File "c:\users\xiaom\desktop\ml-agents\ml-agents\mlagents\trainers\policy.py", line 107, in _execute_model
    network_out = self.sess.run(list(out_dict.values()), feed_dict=feed_dict)
  File "C:\Users\xiaom\Anaconda3\lib\site-packages\tensorflow\python\client\session.py", line 905, in run
    run_metadata_ptr)
  File "C:\Users\xiaom\Anaconda3\lib\site-packages\tensorflow\python\client\session.py", line 1140, in _run
    feed_dict_tensor, options, run_metadata)
  File "C:\Users\xiaom\Anaconda3\lib\site-packages\tensorflow\python\client\session.py", line 1321, in _do_run
    run_metadata)
  File "C:\Users\xiaom\Anaconda3\lib\site-packages\tensorflow\python\client\session.py", line 1340, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: NodeDef expected inputs 'int32_ref, int32' do not match 0 inputs specified; Op<name=Assign; signature=ref:Ref(T), value:T -> output_ref:Ref(T); attr=T:type; attr=validate_shape:bool,default=true; attr=use_locking:bool,default=true; allows_uninitialized_input=true>; NodeDef: global_step/Assign_1 = Assign[T=DT_INT32, _class=["loc:@global_step"], use_locking=true, validate_shape=true]()
INFO:tensorflow:Restoring parameters from ./models/test1-0/3DBallLearning\model-5204.cptk
INFO:tensorflow:Froze 20 variables.
Converted 20 variables to const ops.
INFO:mlagents.envs:Exported .bytes file
ConsoleCtrlHandler function failedException ignored in: <module 'threading' from 'C:\\Users\\xiaom\\Anaconda3\\lib\\threading.py'>
Traceback (most recent call last):
  File "C:\Users\xiaom\Anaconda3\lib\threading.py", line 1288, in _shutdown
    assert tlock is not None
AssertionError:
```

eshvk commented on ml-agents/mlagents/trainers/trainer_controller.py:
```diff
@@ -6,6 +6,8 @@
 import logging
 import shutil
+import sys
+import win32api, win32con
```
Can you make this import conditional? Perhaps condition on whether `os.name` is `nt`? I'd prefer avoiding unnecessary imports.

xiaomaogy replied:
Will change.

eshvk commented on ml-agents/mlagents/trainers/trainer_controller.py:
```diff
@@ -181,12 +184,29 @@ def _save_model(self,steps=0):
         """
         for brain_name in self.trainers.keys():
             self.trainers[brain_name].export_model()
+        self.logger.info("Exported .bytes file")
```
Should this be `self.logger.debug`?

xiaomaogy replied:
I used self.logger.info here because this is to notify our user the .bytes file is exported, similar to what we did in def _save_model(self,steps=0):
where we said `self.logger.info('Saved Model')`

xiaomaogy replied:
Fixed

eshvk commented on ml-agents/mlagents/trainers/trainer_controller.py:
```diff
@@ -288,12 +308,14 @@ def start_learning(self):
             for brain_name, trainer in self.trainers.items():
                 trainer.write_tensorboard_text('Hyperparameters',
                                                trainer.parameters)
+            if sys.platform == 'win32':
```
I am a bit confused here, is `sys.platform` really equal to `win32` even when the underlying OS is 64 bit windows?

xiaomaogy replied:
Yes even if os is 64 bit, sys.platform is win32, see https://stackoverflow.com/questions/2144748/is-it-safe-to-use-sys-platform-win32-check-on-64-bit-python.

eshvk replied:
👍 Thinking a little bit more, let us make this `sys.platform.startswith('win')`, so that we defend against future versions of python where they make a separate `win64` string.

xiaomaogy replied:
Fixed

vincentpierre commented on ml-agents/mlagents/trainers/learn.py:
```diff
@@ -117,3 +117,7 @@ def main():
+
+# For python debugger to directly run this script
+if __name__ == "__main__":
+    main()
```
Add new line

xiaomaogy replied:
Fixed

eshvk commented on ml-agents/mlagents/trainers/trainer_controller.py:
```diff
@@ -6,6 +6,10 @@
 import logging
 import shutil
+import sys
+if sys.platform == 'win32':
```
Do you mind making this `if sys.platform.startswith('win'):`, this will be more future proof, especially if they change this string in future versions of platform. See [here](https://docs.python.org/2/library/sys.html#sys.platform).

xiaomaogy replied:
Fixed

eshvk commented:
👍

xiaomaogy commented:
Now after ctrl-c, it will look like this:

```
  File "C:\Users\ML\Anaconda3\lib\site-packages\tensorflow\python\client\session.py", line 1340, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: NodeDef expected inputs 'int32_ref, int32' do not match 0 inputs specified; Op<name=Assign; signature=ref:Ref(T), value:T -> output_ref:Ref(T); attr=T:type; attr=validate_shape:bool,default=true; attr=use_locking:bool,default=true; allows_uninitialized_input=true>; NodeDef: global_step/Assign_1 = Assign[T=DT_INT32, _class=["loc:@global_step"], use_locking=true, validate_shape=true]()
INFO:tensorflow:Restoring parameters from ./models/test6-1/3DBallLearning\model-1018.cptk
INFO:tensorflow:Restoring parameters from ./models/test6-0/3DBallLearning\model-2098.cptk
INFO:tensorflow:Froze 20 variables.
Converted 20 variables to const ops.
INFO:mlagents.trainers:Exported 3DBallLearning.bytes file
INFO:mlagents.envs:-------------Your model is exported correctly, please ignore the InvalidArgumentError and the AssertionError-------------
ConsoleCtrlHandler function failedException ignored in: <module 'threading' from 'C:\\Users\\ML\\Anaconda3\\lib\\threading.py'>
Traceback (most recent call last):
  File "C:\Users\ML\Anaconda3\lib\threading.py", line 1288, in _shutdown
    assert tlock is not None
AssertionError:
INFO:tensorflow:Froze 20 variables.
Converted 20 variables to const ops.
INFO:mlagents.trainers:Exported 3DBallLearning.bytes file
INFO:mlagents.envs:-------------Your model is exported correctly, please ignore the InvalidArgumentError and the AssertionError-------------
ConsoleCtrlHandler function failedException ignored in: <module 'threading' from 'C:\\Users\\ML\\Anaconda3\\lib\\threading.py'>
Traceback (most recent call last):
  File "C:\Users\ML\Anaconda3\lib\threading.py", line 1288, in _shutdown
    assert tlock is not None
AssertionError:
```
